metpy/calc/tools.py

```diff
@@ -330,3 +332,142 @@ def log_interp(x, xp, fp, **kwargs):
+
+    # Bottom of layer calculations
+
+    # If no bottom is specified, use the first pressure value
+    if not bottom:
+        bottom_pressure = p[0]
+        if heights is not None:
+            bottom_height = heights[0]
+        else:
+            bottom_height = pressure_to_height_std(bottom_pressure)
+
+    # If the bottom pressure is specified
+    if bottom:
+        # in pressure units, assign it
+        if bottom.dimensionality == {'[length]': -1.0, '[mass]': 1.0, '[time]': -2.0}:
+            bottom_pressure = bottom
+            if heights is not None:
+                bottom_height = log_interp(bottom_pressure, p, heights)
+            else:
+                bottom_height = pressure_to_height_std(bottom_pressure)
+        # in height units
+        if bottom.dimensionality == {'[length]': 1.0}:
+            bottom_height = bottom
+            if heights is not None:
+                bottom_pressure = np.interp(np.array([bottom_height.m]), heights,
+                                            p)[0] * p.units
+            else:
+                bottom_pressure = height_to_pressure_std(bottom_height)
+
+    # Top of layer calculations
+
+    # Depth is in pressure
+    if depth.dimensionality == {'[length]': -1.0, '[mass]': 1.0, '[time]': -2.0}:
+        top_pressure = bottom_pressure - depth
+        if heights is not None:
+            top_height = log_interp(top_pressure, p, heights)
+        else:
+            top_height = pressure_to_height_std(top_pressure)
+
+    # Depth is in height
+    if depth.dimensionality == {'[length]': 1.0}:
+        top_height = bottom_height + depth
+        if heights is not None:
+            top_pressure = np.interp(np.array([top_height.m]), heights, p)[0] * p.units
+        else:
+            top_pressure = height_to_pressure_std(top_height)
+
+    # Handle top or bottom values that are invalid
+    if bottom_pressure > p[0]:
+        raise ValueError(
+            'Bottom of layer pressure is greater than the largest pressure in data.')
+    if top_pressure < p[-1]:
+        raise ValueError('Top of layer pressure is less than the lowest pressure in data.')
+    if bottom_pressure < top_pressure:
+        raise ValueError(
+            'Pressure at the top of the layer is greater than that at the bottom.')
+
+    ret = []  # returned data variables in layer
+
+    # Ensure pressures are sorted in ascending order
+    sort_inds = np.argsort(p)
+    p = p[sort_inds]
+
+    # Mask based on top and bottom pressure
+    inds = (p <= bottom_pressure) & (p >= top_pressure)
+    p_interp = p[inds]
+
+    # Interpolate pressures at bounds if necessary and sort
+    if interpolate:
+        # If we don't have the bottom or top requested, append them
+        if top_pressure not in p_interp:
+            p_interp = np.sort(np.append(p_interp, top_pressure)) * p.units
+        if bottom_pressure not in p_interp:
+            p_interp = np.sort(np.append(p_interp, bottom_pressure)) * p.units
+
```

Comment:
I think you should have one big path for `if interpolate` and an `else` for not. Move `inds` from above into the `else` clause since that's the only time it's used. Seems better to opt to duplicate the loop over `args` and the `append` so that we can reduce branches. You can always do this before the `if`:
```python
args = (datavar[sort_inds] for datavar in args)
```